### PR TITLE
Add support for String and &str types, and calls to to_string on those

### PR DESCRIPTION
**Pull request by romac**
- Extracts both `String` and `&str` as a Stainless `String`.
- Calls to `std::fmt::ToString::to_string` on values of this type are erased to just the underlying value (since the value is extracted in both cases to a value of type `String`).

### CODE REVIEW

File: stainless_extraction/src/expr.rs [hunk 1]
```diff
@@ -363,6 +363,15 @@ impl<'a, 'l, 'tcx> BodyExtractor<'a, 'l, 'tcx> {
...
 
+    if fd_id.symbol_path == ["std", "string", "ToString", "to_string"] {
+      return dbg!(self.extract_expr_ref(args.first().cloned().unwrap()));
+    }
+
+    if fd_id.symbol_path == ["std", "string", "String", "len"] {
+      let string = self.extract_expr_ref(args.first().cloned().unwrap());
+      return self.factory().StringLength(string).into();
+    }
+
```
Comment (romac) [inline stainless_extraction/src/expr.rs:374]: This is a huge hack and should be done properly via `StdItems` instead.

File: stainless_extraction/src/ty.rs [hunk 2]
```diff
@@ -54,11 +54,13 @@ impl<'l, 'tcx> BaseExtractor<'l, 'tcx> {
       TyKind::Int(ast::IntTy::I32) => f.BVType(true, 32).into(),
       TyKind::Int(ast::IntTy::I64) => f.BVType(true, 64).into(),
       TyKind::Int(ast::IntTy::I128) => f.BVType(true, 128).into(),
+      TyKind::Int(ast::IntTy::Isize) => f.BVType(true, 64).into(),
```
Comment (romac) [inline stainless_extraction/src/ty.rs:57]: This should map to something like `BVType(true, std::mem::size_of::<usize>() * 8)`.

File: stainless_extraction/src/ty.rs [hunk 3]
```diff
@@ -54,11 +54,13 @@ impl<'l, 'tcx> BaseExtractor<'l, 'tcx> {
...
       TyKind::Uint(ast::UintTy::U8) => f.BVType(false, 8).into(),
       TyKind::Uint(ast::UintTy::U16) => f.BVType(false, 16).into(),
       TyKind::Uint(ast::UintTy::U32) => f.BVType(false, 32).into(),
       TyKind::Uint(ast::UintTy::U64) => f.BVType(false, 64).into(),
       TyKind::Uint(ast::UintTy::U128) => f.BVType(false, 128).into(),
+      TyKind::Uint(ast::UintTy::Usize) => f.BVType(false, 64).into(),
```
Comment (romac) [inline stainless_extraction/src/ty.rs:63]: This should map to something like `BVType(true, std::mem::size_of::<isize>() * 8)`.

File: stainless_extraction/src/ty.rs [hunk 4]
```diff
@@ -80,6 +82,12 @@ impl<'l, 'tcx> BaseExtractor<'l, 'tcx> {
         self.extract_ty(substitutions.type_at(0), txtcx, span)
       }
 
+      // String type
+      TyKind::Adt(adt_def, _) if self.is_string(adt_def) => f.StringType().into(),
+
+      // String slice, erased to a plain String
+      TyKind::Str => f.StringType().into(),
```
Comment (romac) [inline stainless_extraction/src/ty.rs:89]: Maybe match on a `TyKind::Ref(TyKind::Str)` instead?

File: stainless_extraction/src/expr.rs [hunk 5]
```diff
@@ -387,6 +393,38 @@ impl<'a, 'l, 'tcx> BodyExtractor<'a, 'l, 'tcx> {
...
+  fn extract_box_new(
+    &mut self,
+    fd_id: &st::SymbolIdentifier,
+    args: &Vec<ExprRef<'tcx>>,
+  ) -> Option<st::Expr<'l>> {
+    if fd_id.symbol_path != ["std", "boxed", "Box", "T", "new"] {
+      return None;
+    }
+
+    Some(self.extract_expr_ref(args.first().cloned().unwrap()))
```
Comment (yannbolliger) [inline stainless_extraction/src/expr.rs:405]: I've activated `bool_to_option` which let's you do this:
```rust
(fd_id.symbol_path != ["std", "boxed", "Box", "T", "new"])
  .then(|| self.extract_expr_ref(args.first().cloned().unwrap()))
```

File: stainless_extraction/src/ty.rs (hunk 4 above)
Comment (yannbolliger) [reply stainless_extraction/src/ty.rs:116]: That sounds like a good idea.

File: stainless_extraction/src/expr.rs [hunk 6]
```diff
@@ -352,8 +352,13 @@ impl<'a, 'l, 'tcx> BodyExtractor<'a, 'l, 'tcx> {
     // Special case for Box::new, erase it and return the argument directly.
     // TODO: turn Box::new to a StdItem and use that. Tracked here:
     //  https://github.com/epfl-lara/rust-stainless/issues/34
```
Comment (yannbolliger) [inline stainless_extraction/src/expr.rs:354]: That TODO should probably move to the function.

**Review by yannbolliger — APPROVED** [review]
Very cool 🚀 
Should we rename the #34 issue to include strings and close them together once we refactor std item detection (aka #91 or follow-up)?

File: stainless_frontend/tests/pass/strings.rs [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+  take_string(return_string());
+  take_string(return_static_str().to_string());
+  take_static_str(return_static_str());
+  take_str(return_static_str());
+}
+
+#[allow(unused_variables)]
+#[pre(s.len() > 0)]
+fn need_non_empty(s: &str) {
+  assert!(true)
```
Comment (romac) [inline stainless_frontend/tests/pass/strings.rs:31]:
```suggestion
  assert!(s.len() != 0);
```

**Review by romac — COMMENTED** [review]
Left a quick comment to improve a test, looks good otherwise!

**Pull request merged**